Fix reinitialization issues when using total forces in NAMD

jhenin opened the pull request:
(2025-12-09: rewritten description - @giacomofiorin)

Fixes two bugs triggered by the (relatively infrequent) combination of features when total forces are being used in a simulation where the system's topology is changed mid-run:
- Fix a boundary error when accessing atomic total forces from the previous step
- Fix wrong computation of the total forces after a configuration change (the symptom was a `nan` in the output, due to a division by zero because not all Colvars properties had been set previously)

Fixes #778

giacomofiorin committed:
cleanup: Remove unused flag

giacomofiorin commented:
FYI this is also in the integrate_cudagm branch

jhenin commented:
Ah I did not see that branch without a PR. If you'd rather merge the other branch soon, please close this PR and open a PR with `integrate_cudagm`.

HanatoK committed:
fix: try fixing the crash

Node.h is needed for macros like `CMK_SMP && USE_CKLOOP`, and
"NamdTypes.h" is used for NAMD_UNIFIED_REDUCTION and AtomIDList.

HanatoK commented:
While commit https://github.com/Colvars/colvars/pull/887/commits/8cc8ef9c65fa37820e8bf3130e059fc3af948896 seems to fix the problem, in local testing I still get a crash from `003_reinitatoms` since `atoms_map` is not correctly re-initialized after `cv reset`.

giacomofiorin commented:
>     * When initializing atom groups, Colvars has to communicate with the proxy directly (the SOA code `init_atom_from_proxy` inherits the design from the AOS code). In the case of `colvarproxy_namd`, this requires that the `atoms_map` to be initialized with -1, but after `cv reset` the map is cleared. There is no way to re-initialize `atoms_map` after `colvarproxy_namd::reset()` since `colvarproxy::add_config` and `colvarproxy::engine_ready()` are not overridable. The only hacky way is to call `init_atoms_map()` in `colvarproxy_namd::reset()`.

I would like to have a base-class function for the atoms map, because it's not a NAMD-specific issue.  LAMMPS and GROMACS are not using it yet, but simply because they lack this optimization.

>     * I don't know why the atom IDs with total forces enabled in NAMD GM are not filtered out for each client. I could apply the following patch to solve the issue:

Ask the other folks at UIUC (including Jim), and you will find no fans of GlobalMaster's multiplexed design. But the alternative would have been multiple rounds of reduction/broadcast, and I think at the time it was not only more coding work but also potentially harmful to multinode performance.

> (ii) I don't know if Colvars needs to "introspect" the forces from TCL forces. In other words, I don't know if this is what Colvars want;

I don't think we know what each person using Colvars really wants, but I agree that that's a very unlikely use case. 

> @giacomofiorin Could you look into the issues above?

Well, you did that just hours after

giacomofiorin commented:
> I'm not sure. Maybe you can make any one of `add_config`, `engine_ready` and `parse_module_config` virtual and re-implement it with calling `init_atoms_map` before calling the corresponding base class function.

Yeah, that's definitely required, I'll do that.

 The flip side of it is catching the use case when the system's topology changes while Colvars is still defined. That may need a change in NAMD.

giacomofiorin committed:
opt: Do not invalidate NAMD GlobalMaster arrays when not needed

Now that we are overriding the GlobalMaster class functions, we can pass protected members
through a const accessor

giacomofiorin committed:
fix: Update atoms_map in NAMD proxy after requesting new atoms

HanatoK committed:
test: force the boundary checks to reveal the bug

This is just to trigger the bug on CI. You can revert this commit later
after solving the issue.

HanatoK committed:
fix: reset the atoms_map to -1 instead of clearing it

This might to solve the resetting and reloading crash in the NAMD
interface.

HanatoK committed:
test: try the 003_reinitatoms test on ARM64

HanatoK committed:
test: remove the skip_test.sh and try the test with ARM64 again

HanatoK committed:
test: revert the previous commits to disable ARM64 test again

Currently spiff has no way to ignore the comparison of nan and -nan.

giacomofiorin committed:
fix: Track when total forces from the engine are valid or not

giacomofiorin committed:
test: Update regtest reference file to remove nan

giacomofiorin committed:
test: Re-enable 003_reinitatoms test in ARM64

giacomofiorin committed:
doc: Set bug error flag upon bug occurrence

giacomofiorin committed:
fix: Backport changes from LAMMPS repository

giacomofiorin committed:
test: Try downloading container a couple of times before raising error

giacomofiorin committed:
fix: Fix container name in CI job

giacomofiorin committed:
fix: Access colvarproxy_lammps object from correct rank

giacomofiorin commented:
@jhenin I think this is done now. Take a look?

giacomofiorin commented on src/colvar.cpp:
```diff
@@ -1644,6 +1644,16 @@ int colvar::collect_cvc_gradients()
+    ft.reset();
+    return COLVARS_OK;
+  }
+
```
This is expressing the condition that they would not be expected, because they are not valid.

It is my best-effort attempt to generalize the `step_relative() > 0` condition: I'd be open to other ideas, of course.

HanatoK committed:
fix: skip the total force calc in the GPU code path if unavailable

HanatoK reviewed (changes requested):
Looks good to me overall. Can you take a look at the GPU code path in https://github.com/Colvars/colvars/pull/887/commits/ddd8277151b4a65122c67135a0a357289297a6b6 and provide a way to reset the protected `ft` field there?

giacomofiorin committed:
refactor: Invalidate colvar total force from a function used also by GPU scheduler

HanatoK commented on src/colvar.cpp:
```diff
@@ -1407,6 +1407,10 @@ int colvar::calc()
     error_code |= calc_cvcs();
+    if (!cvm::main()->proxy->total_forces_valid()) {
+      // Zero out the colvar total force when atomic total forces are not available
+      ft.reset();
```
My impression is that the GPU code doesn't use any of `colvar::calc` and `colvar::calc_cvcs`. The entry point of the GPU code is in `colvarmodule.cpp`:
https://github.com/Colvars/colvars/blob/9f72f0d677f1026d8ad97d302f33c893868296fd/src/colvarmodule.cpp#L1091-L1093
The idea of the GPU code there is flattening the whole tree-like structure in Colvars, constructing a single CUDA graphs from sub-graphs of all atom groups, and then computing all CVCs and their total forces. In other words, the serial code traverses the `colvardeps*` tree by post-order, while the GPU code traverses it by something similar to level-order.

giacomofiorin committed:
refactor: Define accessor to reset the colvar total force

giacomofiorin committed:
fix: Apply consistent check between total-force compute / collect steps

HanatoK reviewed (approved):
Looks good to me with the latest commit!

giacomofiorin committed:
refactor: When atomic total forces are invalid, ensure colvar total force is zero after collection

giacomofiorin commented on src/colvar_gpu_calc.cpp:
```diff
@@ -357,11 +360,16 @@ int colvarmodule_gpu_calc::cvc_calc_Jacobian_derivative(const std::vector<colvar
 #endif // defined (COLVARS_NVTX_PROFILING)
   for (auto cvi = colvars.begin(); cvi != colvars.end(); cvi++) {
-    error_code |= (*cvi)->collect_cvc_data();
+    if (total_force_valid) {
```
Yeah, but it doesn't hurt to have for clarity I think.

giacomofiorin commented on src/colvar.h:
```diff
@@ -82,6 +82,12 @@ class colvar : public colvarparse, public colvardeps {
+  /// Reset the total force to zero
+  inline void reset_total_force()
+  {
+    ft.reset();
```
Good point, `ft_reported` should still be set based on `ft` but let's make it failproof...

giacomofiorin committed:
fix: Ensure that reported colvar total force is zeroed immediately

HanatoK committed:
fix: ensure that total forces are reset before calc_colvar_properties if
invalid

This commit simplifies the logic of resetting the total forces by
just checking if they are available before accumulating them in
collect_cvc_total_forces. This can ensure that ft_reported is not reset
after calc_colvar_properties. This commit also simplifies the GPU code
path by just reusing the check in collect_cvc_total_forces.

HanatoK committed:
fix: fix the typos in previous rebase

HanatoK commented on src/colvarproxy.cpp:
```diff
@@ -613,6 +609,21 @@ int colvarproxy::post_run()
+void colvarproxy::set_total_forces_invalid()
+{
+  std::fill(atoms_total_forces.begin(), atoms_total_forces.end(), cvm::rvector(0.0, 0.0, 0.0));
+  std::fill(atom_groups_total_forces.begin(), atom_groups_total_forces.end(),
```
I'm not sure if it's necessary to clear the buffers. It looks fine even if the GPU buffers are not cleared.

HanatoK replied:
Either clearing or not is OK. I believe `set_total_forces_invalid` is not called very frequently so the performance is an issue.
